src/services/sourcemaps.ts

```diff
@@ -0,0 +1,330 @@
+            getGeneratedPosition
+        };
+
+        function getGeneratedPosition(loc: SourceMappableLocation): SourceMappableLocation {
+            const maps = filter(getForwardSortedMappings(), m => comparePaths(loc.fileName, m.sourcePath, sourceRoot) === 0);
+            if (!length(maps)) return loc;
+            let targetIndex = binarySearch(maps, { sourcePosition: loc.position }, getSourcePosition, compareValues);
+            if (targetIndex < 0 && maps.length > 0) {
+                // if no exact match, closest is 2's compliment of result
+                targetIndex = ~targetIndex;
+            }
+            return { fileName: toPath(map.file, sourceRoot, host.getCanonicalFileName), position: maps[targetIndex].emittedPosition }; // Closest span
+        }
+
+        function getOriginalPosition(loc: SourceMappableLocation): SourceMappableLocation {
+            const maps = getReverseSortedMappings();
+            if (!length(maps)) return loc;
+            let targetIndex = binarySearch(maps, { emittedPosition: loc.position }, identity, compareProcessedSpanEmittedPositions);
+            if (targetIndex < 0 && maps.length > 0) {
+                // if no exact match, closest is 2's compliment of result
+                targetIndex = ~targetIndex;
+            }
+            return { fileName: toPath(maps[targetIndex].sourcePath, sourceRoot, host.getCanonicalFileName), position: maps[targetIndex].sourcePosition }; // Closest span
+        }
+
+        function getSourceFileLike(fileName: string, location: string): SourceFileLike | undefined {
+            // Lookup file in program, if provided
+            const file: SourceFileLike = program && program.getSourceFile(fileName);
+            if (!file) {
+                // Otherwise check the cache (which may hit disk)
+                const path = toPath(fileName, location, host.getCanonicalFileName);
+                return fallbackCache.get(path);
+            }
+            return file;
+        }
+
+        function getPositionOfLineAndCharacterUsingName(fileName: string, directory: string, line: number, character: number) {
+            const file = getSourceFileLike(fileName, directory);
+            if (!file) {
+                return -1;
+            }
+            return getPositionOfLineAndCharacter(file, line, character);
+        }
+
+        function getDecodedMappings() {
+            return decodedMappings || (decodedMappings = calculateDecodedMappings());
+        }
+
+        function getReverseSortedMappings() {
+            return reverseSortedMappings || (reverseSortedMappings = getDecodedMappings().slice().sort(compareProcessedSpanSourcePositions));
+        }
+
+        function getForwardSortedMappings() {
+            return forwardSortedMappings || (forwardSortedMappings = getDecodedMappings().slice().sort(compareProcessedSpanEmittedPositions));
+        }
+
+        function calculateDecodedMappings(): ProcessedSourceMapSpan[] {

```

Comment:
This could probably be done without so many fields on `state` if you inline `hasCompletedDecoding` and `decodeSingleSpan`. It doesn't need to be done in this PR but it seems unnecessarily complex.